player_linux.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2017 The Oto Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+package oto
+
+// #cgo LDFLAGS: -lasound
+// #include <alsa/asoundlib.h>
+// #include "player_linux.h"
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type player struct {
+	handle     *C.snd_pcm_t
+	buf        []byte
+	bufSamples int
+}
+
+func alsaError(err C.int) error {
+	return fmt.Errorf("oto: ALSA error: %s", C.GoString(C.snd_strerror(err)))
+}
+
+func newPlayer(sampleRate, numChans, bytesPerSample, bufferSizeInBytes int) (*player, error) {
+	var p player
+
+	// open a default ALSA audio device for blocking stream playback
+	errCode := C.snd_pcm_open(&p.handle, C.CString("default"), C.SND_PCM_STREAM_PLAYBACK, 0)
+	if errCode < 0 {
+		return nil, alsaError(errCode)
+	}
+
+	// bufferSize is the total size of the main circular buffer fullness of this buffer
+	// oscilates somewhere between bufferSize and bufferSize-periodSize
+	bufferSize := C.snd_pcm_uframes_t(bufferSizeInBytes / (numChans * bytesPerSample))
+	// periodSize is the number of samples that will be taken from the main circular
+	// buffer at once, we leave this value to bufferSize, because ALSA will change that
+	// to the maximum viable number, obviously lower than bufferSize
+	periodSize := bufferSize
+
+	// choose the correct sample format according to bytesPerSamples
+	var format C.snd_pcm_format_t
+	switch bytesPerSample {
+	case 1:
+		format = C.SND_PCM_FORMAT_S8
+	case 2:
+		format = C.SND_PCM_FORMAT_S16_LE
+	default:
+		panic(fmt.Errorf("oto: bytesPerSample can be 1 or 2, got %d", bytesPerSample))
```

Comment:
Fixed, although I suggest this should be handled at the main platform-independent constructor.